### ISSUE TITLE
[BUG] gdb and gdbserver are out of date 

### ISSUE THREAD

**Issue by dhilburn** [android/ndk#1778]
### Description

I'm using Visual Studio's Android support, which relies on Android NDK build and debug tools. I'm currently seeing an issue where Visual Studio's debugger will crash when hitting a breakpoint on a non-UI thread (the UI thread works just fine). The reason is that it encounters an unhandled-command error from gdbserver: "Call Frame Instruction op 45 in vendor extension space is not handled on this architecture."

I originally filed a bug with Visual Studio support (https://developercommunity.visualstudio.com/t/VS-2022-Preview-C-Android-debugger-cra/10165250?space=62&entry=problem), who sent me over to Sourceware support (https://sourceware.org/bugzilla/show_bug.cgi?id=29681), who diagnosed the error as my gdb version being out-of-date. It looks like Visual Studio is simply deploying the gdbserver version from the NDK, so now I'm here to ask if you guys can update that version.

I'm currently using NDK 25.1.8937393, although that version doesn't seem to contain any prebuilt gdbserver app files. Visual Studio is currently getting those from NDK 23.1.7779620, which I also have installed.  Visual Studio seems to be using adb to communicate with gdbserver, although I'll admit that I'm not terribly familiar yet with specifically how all that works, so I could be mistaken.

Incidentally, tried reproducing this issue in Android Studio, but it worked just fine. The debug output window seemed to indicate that Android Studio was using lldbserver instead of gdbserver. Is that correct? If so, does that mean that the NDK is dropping support for gdbserver? If so, I can request that Visual Studio update their Android support package to use that instead.

Thanks so much for any information you can give me on this issue. Given that native Android apps need to run on a non-UI thread, this bug makes it terribly difficult to get my game engine code ported to Android. =D

Incidentally, I have a simple test app to demonstrate the issue if that is helpful for you guys. Let me know, and I'll attach the .zip file to this issue.

### Affected versions

r25

### Host OS

Windows

### Host OS version

Windows 11

### Affected ABIs

arm64-v8a

### Build system

Other (specify below)

### Other build system

Visual Studio 22 Preview

### minSdkVersion

31

### Device API level

31

**Comment by dhilburn** [comment]
Ah, thanks so much for the help. I've updated my Visual Studio support issue to request that they update their tools to use lldb or to distribute their own gdb+gdbserver builds for Android.

In the meanwhile, can you point me to any resources for download or building an updated gdb+gdbserver for Android on Windows? What process were you guys using?

**Comment by enh-google** [comment]
> Ah, thanks so much for the help. I've updated my Visual Studio support issue to request that they update their tools to use lldb or to distribute their own gdb+gdbserver builds for Android.

thanks! (given how much IDE developers seem to prefer lldb, i'd assume they also have lldb support ready and waiting, and only used gdb for Android for historical reasons?)

> In the meanwhile, can you point me to any resources for download or building an updated gdb+gdbserver for Android on Windows? What process were you guys using?

on Windows? we don't even build the Windows clang _on_ Windows ... we only ever cross-compile from Linux.

additionally, we were never able to get upstream gdb to take our Android support patches. iirc, it doesn't even build out of the box with a current NDK, and even if you get past that, it's missing support for the domain sockets stuff that we used to be able to debug despite the various security restrictions.

and even if you get past those hurdles (which, though non-trivial, is _doable_; if you're brave enough, you can reverse engineer our last patches by diffing the NDK r23 source against the corresponding gdb version, and then port those to the current gdb) ... one thing we learned in the course of our llvm transition is that -- although things like DWARF are theoretically standardized -- in practice, a compiler and its associated tools co-evolve to only really support each other. specifically for your case, we had mixed results debugging clang-built code with gdb (or gcc-built code with lldb); sometimes it works, sometimes it fails obviously, sometimes it fails subtly.

TL;DR: you _could_ try this, but it's not for the faint of heart! (iirc the last gdb update i did took me a couple of weeks to get something that mostly worked, and i already had the pieces lying around from the previous update :-( )

**Comment by pirama-arumuga-nainar** [comment]
[AGDE](https://developer.android.com/games/agde) is aimed at game developers and appears to support lldb.  Can you give that a try?

**Comment by dhilburn** [comment]
> [AGDE](https://developer.android.com/games/agde) is aimed at game developers and appears to support lldb. Can you give that a try?

😮 I had no idea that existed! Thanks so much! I'll definitely give that a try!

**Comment by enh-google** [comment]
don't forget this bit...

> @DanAlbert --- i notice that the roadmap (https://android.googlesource.com/platform/ndk/+/master/docs/Roadmap.md#historical-releases) doesn't actually mention r24 at all, despite it being a release that contained several big changes like this...

**Comment by DanAlbert** [comment]
Doesn't seem related to the bug, so I closed it. I did also send that CL though :)

**Comment by enh-google** [comment]
> Doesn't seem related to the bug, so I closed it. I did also send that CL though :)

sorry for underestimating you --- if _i_ don't have a bug open to remind me, i never know whether i'll ever actually get round to a thing :-(

**Comment by DanAlbert** [comment]
I'm worse than that. This bug wouldn't have helped if I'd forgotten because that task wasn't what was in the title or OP :)

**Comment by dhilburn** [comment]
After a bit of trouble, I was able to get AGDE to kind of work. It seems much better than the Visual Studio Android support, so I'm going to try it out. Most of the AGDE samples don't want to build for me (gradle errors), but EndlessTunnel does work, so I can take it from there. I can confirm that I can put breakpoints on non-UI threads with its lldb debugger, so this issue at least is resolved. Thanks again so much for the link and the information. You guys were increedibly helpful. =)

**Comment by enh-google** [comment]
> Most of the AGDE samples don't want to build for me (gradle errors)

can you include an example so we can pass that on? (or just file a bug via https://source.android.com/docs/setup/contribute/report-bugs#developer-tools [probably the Studio or C++ options] and let us know the url so we can make sure it gets to the right people.)

> Thanks again so much for the link and the information.

(yeah, thanks pirama --- i read "Visual Studio" as "Visual Studio Code" so i could never have made such a useful suggestion!)

dhilburn: was there somewhere in our docs that you looked for help first? if so, we should probably try to add a link advertising ADGE there to help future seekers. (though thanks to you filing a bug with Visual Studio, maybe future seekers will have _multiple_ options, so thanks for that!)

**Comment by DanAlbert** [comment]
(I've reached out to the ADGE owner to figure out where bugs are supposed to go, because I can't find it anywhere on https://developer.android.com/games/agde)

**Comment by dhilburn** [comment]
Sorry for the delay. I got swamped at work and haven't had time to respond. 

>can you include an example so we can pass that on?

It turns out that it was sort of user error. I was expecting the samples to build out-of-the-box, but I don't think that is the intention. The samples page (https://developer.android.com/games/agde/samples) has instructions to get each sample working. Following the directions got all of them to work except for Teapot. Teapot gives the following build error:


Additionally, I did have some trouble at first setting up Java. The quickstart page doesn't really mention Java (https://developer.android.com/games/agde/quickstart), but I got an error message about it when I tried to open the SDK Manager from the Visual Studio toolbar. I downloaded Java 19 (https://www.oracle.com/java/technologies/downloads/#java19) and that silenced the error message, but caused a Java exception instead when launching the SDK Manager. The original error message indicated that I needed to install at least Java 11, so I uninstalled Java 19 and installed Java SE 11 and that got it working.

>was there somewhere in our docs that you looked for help first?

Unfortunately, no. I was starting from the Visual Studio side of things rather than the Android side of things. I actually started incorporating Android into an existing codebase built using msbuild in Visual Studio. I originally set up Windows and later added Linux support using the provided VS Linux example projects which used the new Windows Subsystem for Linux. I planned to do the same thing for Android. When I started, they used Ant to build and package an Android package in a single project. However, in the latest update, they dropped Ant support in favor of Gradle and (as far as I knew) it required a separate project to hold the C++ code, which didn't work for my project structure. I worked to figure out how to build a native Android package manually using msbuild, and got that working using NativeActivity. The primary wrinkle that didn't work was that gdb error which started this whole issue. I had downloaded Android Studio, but was only using it as a reference to compare my Visual Studio work against. Ideally, Visual Studio's Android support would just have been AGDE from the get-go. I'm hoping that they will move to that in a near-future update.

**Comment by DanAlbert** [comment]
> Ideally, Visual Studio's Android support would just have been AGDE from the get-go.

Just to make sure you're not expecting more than what it is, it's not at all _intended_ to be a solution for building typical Android apps. There are no plans for it to support Java or Kotlin. It's supposedly (I haven't used it) a good way to make a game if you're doing so in C++, but it's unlikely to be good for anything else.

**Comment by dhilburn** [comment]
Ah, that makes sense. I'm building a game, so it makes sense that I'd incorrectly see it as the "right" setup.

**Comment by DanAlbert** [comment]
Ah, you're exactly who should be using it then :) Carry on.

**Comment by leozzyzheng** [comment]
I'm also getting error "Call Frame Instruction op 45 in vendor extension space is not handled on this architecture." with ndk r21.
In my case, built with ndk r21 and use ndk r19 gdb works for me although it's not very stable and may stuck sometimes, just give this information for guys still using gdb for debugging.
